File: src/php/Generic/MemoryBenchmarkBase.php
<?php

namespace EntityFX\NetBenchmark\Core\Generic {

    use EntityFX\NetBenchmark\Core\Writer;
    
    abstract class MemoryBenchmarkBase  extends BenchmarkBase {

        public function __construct($writer, $printToConsole)
        {
            parent::__construct($writer, $printToConsole);
            $this->Iterrations = 500000;
            $this->Ratio = 50;
        }

        public function Warmup($aspect = 0.05) {
            $this->UseConsole(DEBUG);
            parent::Warmup($aspect);
            $this->UseConsole(true);
        }

        protected function BenchRandomMemory($aspect = 0.05) {
            $int4k = $this->MeasureArraySpeed(1024);
            $this->output->Write("int 4k: %.2f MB/s", $int4k["MbPerSec"]);
            $this->output->WriteNewLine();
            $int512k = $this->MeasureArraySpeed(131072);
            $this->output->Write("int 512k: %.2f MB/s", $int512k["MbPerSec"]);
            $this->output->WriteNewLine();
            $int8m = $this->MeasureArraySpeed(2097152);
            $this->output->Write("int 8M: %.2f MB/s", $int8m["MbPerSec"]);
            $this->output->WriteNewLine();
            $int32m = $this->MeasureArraySpeed(32 * 1024 * 1024 / 8);
            $this->output->Write("int 32M: %.2f MB/s", $int32m["MbPerSec"]);
            $this->output->WriteNewLine();

            $long4k = $this->MeasureArraySpeedLong(512);
            $this->output->Write("long 4k: %.2f MB/s", $long4k["MbPerSec"]);
            $this->output->WriteNewLine();
            $long512k = $this->MeasureArraySpeedLong(65536);
            $this->output->Write("long 512k: %.2f MB/s", $long512k["MbPerSec"]);
            $this->output->WriteNewLine();
            $long8m = $this->MeasureArraySpeedLong(1048576);
            $this->output->Write("long 8M: %.2f MB/s", $long8m["MbPerSec"]);
            $this->output->WriteNewLine();
            $long32m = $this->MeasureArraySpeedLong(32 * 1024 * 1024 / 8);
            $this->output->Write("long 32M: %.2f MB/s", $long32m["MbPerSec"]);
            $this->output->WriteNewLine();

            $results = [
                $int4k["MbPerSec"],  $int512k["MbPerSec"],   $int8m["MbPerSec"], $int32m["MbPerSec"] ,
                $long4k["MbPerSec"], $long512k["MbPerSec"],  $long8m["MbPerSec"], $long32m["MbPerSec"] ];
            $avg = array_sum($results) / count($results);
            $this->output->Write("Average: %.2f MB/s", $avg);
            $this->output->WriteNewLine();

            return [
                "Average" => $avg,
                "Output" => ""
            ];
        }

        protected function MeasureArraySpeed($size)
        {
            $blockSize = 8;
            $L = [];

            $array = [];
            
            for ($i = 0; $i < $size; $i++) {
                $array[$i] = mt_rand(-2147483648, 2147483647);
            }

            $end = count($array) - 1;
            $k0 = intdiv($size, 1024); 
            $k1 = $k0 == 0 ? 1 : $k0 ;
            $iterInternal = intdiv($this->Iterrations, $k1);
            $iterInternal = $iterInternal == 0 ? 1: $iterInternal;
            for ($idx = 0; $idx < $end; $idx+=$blockSize)
            {
                $L[0] = $array[$idx];
                $L[1] = $array[$idx+1];
                $L[2] = $array[$idx+2];
                $L[3] = $array[$idx+3];
                $L[4] = $array[$idx+4];
                $L[5] = $array[$idx+5];
                $L[6] = $array[$idx+6];
                $L[7] = $array[$idx+7];
            }
            $start = microtime(true);
            for ($i = 0; $i < $iterInternal; $i++)
            {
                for ($idx = 0; $idx < $end; $idx+=$blockSize)
                {
                    $L[0] = $array[$idx];
                    $L[1] = $array[$idx+1];
                    $L[2] = $array[$idx+2];
                    $L[3] = $array[$idx+3];
                    $L[4] = $array[$idx+4];
                    $L[5] = $array[$idx+5];
                    $L[6] = $array[$idx+6];
                    $L[7] = $array[$idx+7];
                }
            }
            $elapsed = microtime(true) - $start;
            return [
                "MbPerSec" => $iterInternal * count($array) * PHP_INT_SIZE / $elapsed / 1024 / 1024,
                "Data" => $L
            ];
        }

        protected function MeasureArraySpeedLong($size)
        {
            $blockSize = 16;
            $L = [];

            $array = [];
            
            for ($i = 0; $i < $size; $i++) {
                $array[$i] = mt_rand(-2147483648, 2147483647);
            }

            $end = count($array) - 1;
            $k0 = intdiv($size, 1024); 
            $k1 = $k0 == 0 ? 1 : $k0 ;
            $iterInternal = intdiv($this->Iterrations, $k1);
            $iterInternal = $iterInternal == 0 ? 1: $iterInternal;
            for ($idx = 0; $idx < $end; $idx+=$blockSize)
            {
                $L[0] = $array[$idx];
                $L[1] = $array[$idx+1];
                $L[2] = $array[$idx+2];
                $L[3] = $array[$idx+3];
                $L[4] = $array[$idx+4];
                $L[5] = $array[$idx+5];
                $L[6] = $array[$idx+6];
                $L[7] = $array[$idx+7];
                $L[8] = $array[$idx+8];
                $L[9] = $array[$idx+9];
                $L[0xA] = $array[$idx+0xA];
                $L[0xB] = $array[$idx+0xB];
                $L[0xC] = $array[$idx+0xC];
                $L[0xD] = $array[$idx+0xD];
                $L[0xE] = $array[$idx+0xE];
                $L[0xF] = $array[$idx+0xF];
            }
            $start = microtime(true);
            for ($i = 0; $i < $iterInternal; $i++)
            {
                for ($idx = 0; $idx < $end; $idx+=$blockSize)
                {
                    $L[0] = $array[$idx];
                    $L[1] = $array[$idx+1];
                    $L[2] = $array[$idx+2];
                    $L[3] = $array[$idx+3];
                    $L[4] = $array[$idx+4];
                    $L[5] = $array[$idx+5];
                    $L[6] = $array[$idx+6];
                    $L[7] = $array[$idx+7];
                    $L[8] = $array[$idx+8];
                    $L[9] = $array[$idx+9];
                    $L[0xA] = $array[$idx+0xA];
                    $L[0xB] = $array[$idx+0xB];
                    $L[0xC] = $array[$idx+0xC];
                    $L[0xD] = $array[$idx+0xD];
                    $L[0xE] = $array[$idx+0xE];
                    $L[0xF] = $array[$idx+0xF];
                }
            }
            $elapsed = microtime(true) - $start;
            return [
                "MbPerSec" => $iterInternal * count($array) * PHP_INT_SIZE / $elapsed / 1024 / 1024,
                "Data" => $L
            ];
        }
    }
}

File: src/lua/generic/memoryBenchmarkBase.lua
MemoryBenchmarkBase = class(BenchmarkBase, function(a, writer, printToConsole)
    BenchmarkBase.init(a, writer, printToConsole)
    a.iterrations = 500000
    a.ratio = 1
end)

function MemoryBenchmarkBase:benchRandomMemory()
    local int4k = self:measureArrayRandomRead(1024)
    self.output:writeLine("int 4k: %.2f MB/s", int4k.mbPerSec)
    local int512k = self:measureArrayRandomRead(131072)
    self.output:writeLine("int 512k: %.2f MB/s", int512k.mbPerSec)
    local int8m = self:measureArrayRandomRead(2097152)
    self.output:writeLine("int 8M: %.2f MB/s", int8m.mbPerSec)
    local int32m = self:measureArrayRandomRead(32 * 1024 * 1024 / 4)
    self.output:writeLine("int 32M: %.2f MB/s", int32m.mbPerSec)

    local long4k = self:measureArrayRandomLongRead(1024)
    self.output:writeLine("long 4k: %.2f MB/s", long4k.mbPerSec)
    local long512k = self:measureArrayRandomLongRead(131072)
    self.output:writeLine("long 512k: %.2f MB/s", long512k.mbPerSec)
    local long8m = self:measureArrayRandomLongRead(2097152)
    self.output:writeLine("long 8M: %.2f MB/s", long8m.mbPerSec)
    local long32m = self:measureArrayRandomLongRead(32 * 1024 * 1024 / 8)
    self.output:writeLine("long 32M: %.2f MB/s", long32m.mbPerSec)
    
    local avg = arithmetic_mean({
        int4k.mbPerSec, int512k.mbPerSec, int8m.mbPerSec, 
        int32m.mbPerSec, long4k.mbPerSec, long512k.mbPerSec,
        long8m.mbPerSec, long32m.mbPerSec})

    self.output:writeLine("Average: %.2f MB/s", avg)

    return {
        average = avg,
        output = self.output.output
    }
end

function MemoryBenchmarkBase:measureArrayRandomRead(size)
    local blockSize = 16
    local I = {}

    local array = randomIntArray(size, 2^31 - 2)

    local endA = #array - 1
    local k0 = (size / 1024)

    local k1 = 1
    if k0 == 0 then k1 = 1 else k1 = k0 end
    local iterInternal = self.iterrations / k1
    if iterInternal == 0 then iterInternal = 1 end
    for idx=0,endA,blockSize do
        I[0] = array[idx]
        I[1] = array[idx + 1]
        I[2] = array[idx + 2]
        I[3] = array[idx + 3]
        I[4] = array[idx + 4]
        I[5] = array[idx + 5]
        I[6] = array[idx + 6]
        I[7] = array[idx + 7]
        I[8] = array[idx + 8]
        I[9] = array[idx + 9]
        I[10] = array[idx + 10]
        I[11] = array[idx + 11]
        I[12] = array[idx + 12]
        I[13] = array[idx + 13]
        I[14] = array[idx + 14]
        I[15] = array[idx + 15]
    end
    local start = os.clock() * 1000
    for i=0,iterInternal-1 do
        for idx=0,endA,blockSize do
            I[0] = array[idx]
            I[1] = array[idx + 1]
            I[2] = array[idx + 2]
            I[3] = array[idx + 3]
            I[4] = array[idx + 4]
            I[5] = array[idx + 5]
            I[6] = array[idx + 6]
            I[7] = array[idx + 7]
            I[8] = array[idx + 8]
            I[9] = array[idx + 9]
            I[10] = array[idx + 10]
            I[11] = array[idx + 11]
            I[12] = array[idx + 12]
            I[13] = array[idx + 13]
            I[14] = array[idx + 14]
            I[15] = array[idx + 15]
        end
    end
    local elapsed = math.floor(os.clock() * 1000 - start)
    return { mbPerSec = (iterInternal * #array * 8.0 / (elapsed / 1000.0) / 1024 / 1024), res = I }
end

function MemoryBenchmarkBase:measureArrayRandomLongRead(size)
    local blockSize = 8
    local I = {}

    local array = randomIntArray(size, 2^31 - 2)

    local endA = #array - 1
    local k0 = (size / 1024)

    local k1 = 1
    if k0 == 0 then k1 = 1 else k1 = k0 end
    local iterInternal = self.iterrations / k1
    if iterInternal == 0 then iterInternal = 1 end
    for idx=0,endA,blockSize do
        I[0] = array[idx]
        I[1] = array[idx + 1]
        I[2] = array[idx + 2]
        I[3] = array[idx + 3]
        I[4] = array[idx + 4]
        I[5] = array[idx + 5]
        I[6] = array[idx + 6]
        I[7] = array[idx + 7]
    end
    local start = os.clock() * 1000
    for i=0,iterInternal-1 do
        for idx=0,endA,blockSize do
            I[0] = array[idx]
            I[1] = array[idx + 1]
            I[2] = array[idx + 2]
            I[3] = array[idx + 3]
            I[4] = array[idx + 4]
            I[5] = array[idx + 5]
            I[6] = array[idx + 6]
            I[7] = array[idx + 7]
        end
    end
    local elapsed = math.floor(os.clock() * 1000 - start)
    return { mbPerSec = (iterInternal * #array * 8.0 / (elapsed / 1000.0) / 1024 / 1024), res = I }
end
